### DISCUSSION TITLE
Latency

### DISCUSSION THREAD

**Discussion by improperaffo** [srsran/srsRAN_Project#234] [General Help]
Dear all,
we achieved decent throughput using srsRAN, but we still have a high latency on the link between the core and the UEs. The best we achieved when pinging a single UE from the UPF is
> --- 10.45.0.17 ping statistics ---
1000 packets transmitted, 1000 received, 0% packet loss, time 20005ms
rtt min/avg/max/mdev = 6.745/14.846/33.874/2.620 ms, pipe 2

We got his results using the following config file and open5gs as core

```
cell_cfg:
  dl_arfcn: 632628
  band: 78
  channel_bandwidth_MHz: 20
  common_scs: 30
  plmn: "00101"
  tac: 1
  pci: 1
  tdd_ul_dl_cfg:
    dl_ul_tx_period: 4
    nof_dl_slots: 2
    nof_ul_slots: 1
  nof_antennas_ul: 1
  nof_antennas_dl: 1
  pdsch:
    mcs_table: qam256
  pusch:
    mcs_table: qam256
  pucch:
    sr_period_ms: 4

expert_phy:
  max_proc_delay: 1
  nof_pdsch_threads: 2
```

Any idea on how to go forward? Also, is there a way in SRS to track a buffer through the stack to understand where the latency is coming from?

**Comment by improperaffo** [comment]
UPDATE:
After running some benchmarks, we discovered that decoding has a big impact on the processing time. So, after limiting the UL MCS to BPSK (not ideal for throughput), we could achieve down to 11 ms average RTT.

Here are the times taken by our CPU to encode and decode symbols:
```
root@srs-ran-gnb-bladerf-szp26:/opt/srsRAN_Project/build/tests/benchmarks/phy/upper/channel_processors# ./pusch_processor_benchmark -m latency
Launching benchmark for 12 threads, 100 times per thread, and 10 repetitions. Using default profile, auto LDPC decoder, and auto rate dematcher.

--- Average latency ---
"PUSCH processor" performance for 10 repetitions. All values are in microseconds.
 Percentiles:                                |  50th  |  75th  |  90th  |  99th  | 99.9th | Worst  |
 PUSCH RB=25  Mod=QPSK   R=0.117 -   0.9 Mbps|    16.6|    16.7|    16.9|    16.9|    16.9|    16.9|
 PUSCH RB=52  Mod=QPSK   R=0.117 -   1.9 Mbps|    29.4|    29.6|    29.8|    29.8|    29.8|    29.8|
 PUSCH RB=106 Mod=QPSK   R=0.117 -   3.8 Mbps|    57.3|    57.5|    67.5|    67.5|    67.5|    67.5|
 PUSCH RB=270 Mod=QPSK   R=0.117 -   9.7 Mbps|   157.1|   172.5|   714.7|   714.7|   714.7|   714.7|
 PUSCH RB=25  Mod=16QAM  R=0.643 -  10.0 Mbps|    43.6|    44.5|    70.8|    70.8|    70.8|    70.8|
 PUSCH RB=52  Mod=16QAM  R=0.643 -  21.0 Mbps|    82.3|    82.6|   108.1|   108.1|   108.1|   108.1|
 PUSCH RB=106 Mod=16QAM  R=0.643 -  42.0 Mbps|   164.3|   164.8|   193.2|   193.2|   193.2|   193.2|
 PUSCH RB=270 Mod=16QAM  R=0.643 - 108.6 Mbps|   427.6|   441.7|   450.3|   450.3|   450.3|   450.3|
 PUSCH RB=25  Mod=64QAM  R=0.853 -  20.0 Mbps|    62.5|    62.9|    84.2|    84.2|    84.2|    84.2|
 PUSCH RB=52  Mod=64QAM  R=0.853 -  41.0 Mbps|   124.4|   125.0|   144.2|   144.2|   144.2|   144.2|
 PUSCH RB=106 Mod=64QAM  R=0.853 -  84.0 Mbps|   249.0|   266.6|   279.4|   279.4|   279.4|   279.4|
 PUSCH RB=270 Mod=64QAM  R=0.853 - 217.1 Mbps|   664.4|   664.9|   669.3|   669.3|   669.3|   669.3|
 PUSCH RB=25  Mod=256QAM R=0.926 -  28.7 Mbps|    80.8|    81.1|   101.4|   101.4|   101.4|   101.4|
 PUSCH RB=52  Mod=256QAM R=0.926 -  60.5 Mbps|   162.5|   163.0|   191.3|   191.3|   191.3|   191.3|
 PUSCH RB=106 Mod=256QAM R=0.926 - 123.0 Mbps|   327.2|   349.3|   355.2|   355.2|   355.2|   355.2|
 PUSCH RB=270 Mod=256QAM R=0.926 - 311.4 Mbps|   842.3|   848.1|   882.9|   882.9|   882.9|   882.9|
root@srs-ran-gnb-bladerf-szp26:/opt/srsRAN_Project/build/tests/benchmarks/phy/upper/channel_processors# ./pdsch_processor_benchmark -m latency
Launching benchmark for 12 threads, 100 times per thread, and 10 repetitions. Using default profile, and auto LDPC encoder.

--- Average latency ---
"PDSCH processor" performance for 10 repetitions. All values are in microseconds.
 Percentiles:                             |  50th  |  75th  |  90th  |  99th  | 99.9th | Worst  |
 PDSCH RB=25  Mod=QPSK   rv=0 -   0.7 Mbps|     2.5|     2.5|     2.6|     2.6|     2.6|     2.6|
 PDSCH RB=52  Mod=QPSK   rv=0 -   1.5 Mbps|     4.2|     4.2|     4.7|     4.7|     4.7|     4.7|
 PDSCH RB=106 Mod=QPSK   rv=0 -   3.1 Mbps|     7.8|     7.8|     8.0|     8.0|     8.0|     8.0|
 PDSCH RB=270 Mod=QPSK   rv=0 -   7.9 Mbps|    19.4|    19.6|    19.8|    19.8|    19.8|    19.8|
 PDSCH RB=25  Mod=16QAM  rv=0 -   8.1 Mbps|     3.6|     3.7|     4.0|     4.0|     4.0|     4.0|
 PDSCH RB=52  Mod=16QAM  rv=0 -  16.9 Mbps|     7.2|     7.4|     7.8|     7.8|     7.8|     7.8|
 PDSCH RB=106 Mod=16QAM  rv=0 -  34.8 Mbps|    13.4|    13.4|    13.7|    13.7|    13.7|    13.7|
 PDSCH RB=270 Mod=16QAM  rv=0 -  88.1 Mbps|    34.1|    34.2|    34.4|    34.4|    34.4|    34.4|
 PDSCH RB=25  Mod=64QAM  rv=0 -  16.1 Mbps|     5.5|     5.5|     5.7|     5.7|     5.7|     5.7|
 PDSCH RB=52  Mod=64QAM  rv=0 -  33.8 Mbps|    10.8|    10.9|    10.9|    10.9|    10.9|    10.9|
 PDSCH RB=106 Mod=64QAM  rv=0 -  67.6 Mbps|    21.0|    21.1|    21.1|    21.1|    21.1|    21.1|
 PDSCH RB=270 Mod=64QAM  rv=0 - 172.2 Mbps|    53.0|    53.2|    56.1|    56.1|    56.1|    56.1|
 PDSCH RB=25  Mod=256QAM rv=0 -  23.6 Mbps|     6.2|     6.3|     6.4|     6.4|     6.4|     6.4|
 PDSCH RB=52  Mod=256QAM rv=0 -  48.2 Mbps|    12.0|    12.0|    12.1|    12.1|    12.1|    12.1|
 PDSCH RB=106 Mod=256QAM rv=0 -  98.4 Mbps|    23.9|    24.2|    53.2|    53.2|    53.2|    53.2|
 PDSCH RB=270 Mod=256QAM rv=0 - 254.2 Mbps|    60.8|    61.5|    91.8|    91.8|    91.8|    91.8|
```

**Reply by ismagom** [reply]
Very interesting results. Thanks for sharing. There are a couple of parameters you could try in the gnb config file: 

```
cell_cfg:
 pucch:
  min_k1: 2 # This k1>=2, you could try even with 1
 pusch:
  min_k2: 2 # This k2>=2, you could try even with 1
```

One additional problem in the uplink processing is that we currently don't apply very high priority to the threads run the processing functions. Maybe you can try increasing the priority [in this line of the code](https://github.com/srsran/srsRAN_Project/blob/5e6f50a202c6efa671d5b231d7c911dc6c3d86ed/apps/gnb/gnb_worker_manager.cpp#L226). 

Thanks again for sharing your results.

**Reply by HaoxinSEU** [reply]
Hi Ismael,

Thanks for these hints. We tried your ideas but unfortunately none of these parameters make further improvement on the latency.

Here are the results.

Reducing the `min_k1` and  `min_k2` values in gNB config file:
```
--- 10.45.0.10 ping statistics ---
5000 packets transmitted, 5000 received, 0% packet loss, time 53990ms
rtt min/avg/max/mdev = 7.404/12.864/38.225/2.645 ms, pipe 4
```

Changing priority to `max() - 1` at line 226 in `gnb_worker_manager.cpp`:
```
--- 10.45.0.4 ping statistics ---
5000 packets transmitted, 5000 received, 0% packet loss, time 51217ms
rtt min/avg/max/mdev = 6.093/12.512/26.034/3.316 ms, pipe 3
```

Changing priority to `max()` at line 226 in `gnb_worker_manager.cpp`:
```
--- 10.45.0.5 ping statistics ---
5000 packets transmitted, 5000 received, 0% packet loss, time 50979ms
rtt min/avg/max/mdev = 7.249/12.562/35.928/3.461 ms, pipe 4
```


And we also have some interesting observations: 
Though setting one thread per CPU core doesn't reduce the latency, but it reduces the variance in latency a lot.
When using AMD Ryzen 7 5800H CPU, which has 8 physical cores, the latency is still around 12 ms, which is similar to our previous results.

Do you have any idea where we can make further improvements? Thanks again

**Reply by pgawlowicz** [reply]
@HaoxinSEU Have you achieved any improvements? Could you report it here?

**Reply by arman-maghsoudnia** [reply]
Hi @pgawlowicz,
We are still working on the topic. We don't have much yet. We will send our results in a different discussion when we get into something.

**Reply by andergalvaan** [reply]
Hi @arman-maghsoudnia @HaoxinSEU 

Are there any updates in how latency can be minimized?

Best regards,
Ander

**Reply by arman-maghsoudnia** [reply]
Hi @andergalvaan,
Reading our latest paper published in HotNets 2024 would give you useful insights about the sources of the latency and the interdependency between them. Here is the link: https://dl.acm.org/doi/10.1145/3696348.3696862
You can manually relate the points made in the paper directly to the srsRAN configurations/implementations. We can discuss more if you have any specific questions about them.

**Reply by fllay** [reply]
Hi @arman-maghsoudnia Thank you for sharing the paper. It is very useful.

**Reply by andergalvaan** [reply]
Hi @arman-maghsoudnia 

Very interesting and helpful paper. I'm currently working on a similar implementation using the same USRP hardware as described in your work. Do you have any implementation available, or could you share srsRAN configuration files? I'm particularly interested in minimizing uplink latency and maximizing uplink throughput.

Best regards,
Ander

**Reply by arman-maghsoudnia** [reply]
Hi @andergalvaan 
We just released our code to achieve low latency with srsRAN here: https://github.com/aygong/srsRAN_Project_Low_Latency
For more technical details and evaluation results, stay tuned for our upcoming workshop paper to be presented at [OpenRIT6G](https://conferences.sigcomm.org/sigcomm/2025/workshop/openrit6g/) workshop in early September.

**Reply by andergalvan** [reply]
Hi @arman-maghsoudnia, thanks for sharing! I’ve opened an issue on GitHub because I’m having trouble compiling the project from source.

**Comment by asheeshtripathi** [comment]
Hi @andergalvaan, we did some benchmarking work with srsRAN and NI/Ettus Research USRPs (X310, B210, and N310). We discussed a few srsRAN parameters relevant for tuning SDR-based 5G over-the-air performance. The impact of these parameters was demonstrated through OTA throughput measurements (both UL and DL UDP throughput) at different distances.

Here is the paper: https://ieeexplore.ieee.org/abstract/document/10978285

For UL-heavy applications, a UL-heavy TDD pattern can also be used, if operating in TDD bands.

**Reply by andergalvaan** [reply]
Hi @asheeshtripathi, thank you for your reply. The paper you mentioned is interesting, but it focuses mainly on the measurement of DL and UL throughput. However, I am mainly interested in minimizing latency. I am also interested in maintaining UL throughput for video transmission, but latency is the more critical for me. Thank you very much!
